### ISSUE TITLE
Widen analysis package:convert dependency to include 3.x

### ISSUE THREAD

**Issue by jonasfj** [dart-lang/sdk#44678]
Similar to https://github.com/dart-lang/sdk/issues/44577

Is there any reason the constraint can't be `>=2.0.0 <4.0.0`?

https://github.com/dart-lang/sdk/blob/3a0a33f35f637e9017b8ed8d727b3ed15577f698/pkg/analyzer/pubspec.yaml#L14

I suspect we should do similar for: 
 * `args`
 * `cli_util`
 * `pub_semver`
 * `watcher`
 * `convert`

At-least `args` is pretty widely used, and anyone using `package:test` has an indirect dependency on `analyzer`, so they'll be blocked from upgrading these packages.

(disclaimer: I haven't checked how breaking the API changes are)

**Comment by devoncarew** [comment]
If these package versions (the null safe versions I assume) have already been DEP'd into the sdk repo, then analyzer is already compatible with their updated APIs.

**Comment by devoncarew** [comment]
https://github.com/dart-lang/sdk/blob/master/DEPS

From the DEPS:
- args; 2c6a221f45e4e0ef447b5d09101bf8a52e1ccd36 (`2.0.0-nullsafety.0`)
- cli_util; 50cc840b146615899e97b892578848401b2028d5 (`0.3.0-nullsafety.0`)
- convert; 6513985a1b1ea8a0b987fbef699250ce2cdc3cca (`3.0.0-nullsafety.0`)
- pub_semver; 10569a1e867e909cf5db201f73118020453e5db8 (`2.0.0-nullsafety.0`)
- watcher; 14706140b3670ca1dfeca73480bdd3091b728ce9 (`1.0.0-nullsafety.0`)

@scheglov - can concerns with reving the constraints to include more null safe deps, and publishing, even before we release a null safe analyzer version? It might make version solves for the ecosystem easier (for me, it would unblock publishing a null safe version of `grinder`, which would then unblock `linter`).

**Comment by scheglov** [comment]
I think there is an issue.
Packages `analyzer` and `_fe_analyzer_shared` are tightly bounded, we publish them only together.

And `_fe_analyzer_shared` is null safe now, requires `sdk: '>=2.12.0-0 <3.0.0'`.
There is no version for SDK before `2.12`.

So, effectively any published version of `analyzer` will require `sdk: '>=2.12.0-0 <3.0.0'`.

If we decide to publish `analyzer` that supports only SDK `2.12`, we could just switch it to specific null safe version of its dependencies, not just widen version ranges. Specifically:
```yaml
dependencies:
  _fe_analyzer_shared: ^14.0.0
  args: ^2.0.0-nullsafety.0
  cli_util: ^0.3.0-nullsafety.0
  collection: ^1.15.0-nullsafety.5
  convert: ^3.0.0-nullsafety.0
  crypto: ^3.0.0-nullsafety.0
  glob: ^2.0.0-nullsafety.0
  meta: ^1.3.0-nullsafety.6
  package_config: ^1.0.0
  path: ^1.8.0-nullsafety.3
  pub_semver: ^2.0.0-nullsafety.0
  source_span: ^1.8.0-nullsafety.4
  watcher: ^0.9.6
  yaml: ^3.0.0-nullsafety.0
dev_dependencies:
  analyzer_utilities:
    path: ../analyzer_utilities
  async: ^2.5.0-nullsafety.3
  linter: any
  matcher: ^0.12.10-nullsafety.3
  pedantic: ^1.10.0-nullsafety.3
  test: ^1.16.0-nullsafety.13
  test_api: ^0.2.19-nullsafety.6
  test_reflective_loader: ^0.2.0-nullsafety.0
```

**Comment by scheglov** [comment]
OTOH, we could publish using an older revision of `analyzer`, which can use the revision of `_fe_analyzer_shared` that was at that moment. The only change would be these widen constraints. I will look for more details.

**Comment by scheglov** [comment]
The CL with changes is here: https://dart-review.googlesource.com/c/sdk/+/179581
It is based on `0.41.1` and only updates `pubspec.yaml`.
It is not for merging, but I need a review.
If it is OK, I will publish it.

I considered restricting `sdk: '>=2.7.0 <2.12.0'`.
But not sure if this is necessary.

Do we want to widen version ranges for any other packages than mentioned in this issue?

**Comment by devoncarew** [comment]
Thanks for the CL! I'll take a quick look.

I don't know how that initial list was determined, but my guess is we'd want to expand to include the null safe version ranges for analyzer's direct, non-dev dependencies.

**Comment by devoncarew** [comment]
In which case, that might just mean expanding the package_config to include 2.0.0? Although package_config is only at `1.9.3` in the dart sdk repo, so perhaps not.

**Comment by devoncarew** [comment]
@scheglov - that CL lgtm.

**Comment by scheglov** [comment]
OK, I will publish it as is then.
If we decide to add more packages, we always can publish another version.

**Comment by scheglov** [comment]
Published `analyzer 0.41.2`.
